Make transpilation setup local to this repo

frangio opened the pull request:
This PR migrates the transpilation of contracts for upgradeability from [the upgadeable repo](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable) to this repo. The result should be a lot simpler to use and understand.

Amxx commented on scripts/upgradeable/upgradeable.patch:
```diff
@@ -0,0 +1,350 @@
+diff --git a/contracts/crosschain/polygon/CrossChainEnabledPolygonChild.sol b/contracts/crosschain/polygon/CrossChainEnabledPolygonChild.sol
+index fa099483..d948b412 100644
+--- a/contracts/crosschain/polygon/CrossChainEnabledPolygonChild.sol
++++ b/contracts/crosschain/polygon/CrossChainEnabledPolygonChild.sol
+@@ -66,7 +66,17 @@ abstract contract CrossChainEnabledPolygonChild is IFxMessageProcessor, CrossCha
+         if (!_isCrossChain()) revert NotCrossChainCall();
+ 
+         _sender = rootMessageSender;
+-        Address.functionDelegateCall(address(this), data, "cross-chain execution failed");
++        __functionDelegateCall(address(this), data);
+         _sender = DEFAULT_SENDER;
+     }
++
++    // ERC1967Upgrade._functionDelegateCall is private so we reproduce it here.
++    // An extra underscore prevents a name clash error.
++    function __functionDelegateCall(address target, bytes memory data) private returns (bytes memory) {
++        require(Address.isContract(target), "Address: delegate call to non-contract");
++
++        // solhint-disable-next-line avoid-low-level-calls
++        (bool success, bytes memory returndata) = target.delegatecall(data);
++        return Address.verifyCallResult(success, returndata, "Address: low-level delegate call failed");
++    }
+ }
```
this should not be necessary anymore

Amxx commented on scripts/upgradeable/upgradeable.patch:
```diff
@@ -0,0 +1,350 @@
+  */
+ abstract contract GovernorVotesComp is Governor {
+     ERC20VotesComp public immutable token;
+diff --git a/contracts/mocks/proxy/UUPSLegacy.sol b/contracts/mocks/proxy/UUPSLegacy.sol
+index 6956f567..6aa10f38 100644
+--- a/contracts/mocks/proxy/UUPSLegacy.sol
++++ b/contracts/mocks/proxy/UUPSLegacy.sol
+@@ -23,7 +23,7 @@ contract UUPSUpgradeableLegacyMock is UUPSUpgradeableMock {
+         // Initial upgrade and setup call
+         __setImplementation(newImplementation);
+         if (data.length > 0 || forceCall) {
+-            Address.functionDelegateCall(newImplementation, data);
++            __functionDelegateCall(newImplementation, data);
+         }
+ 
+         // Perform rollback test if not already in progress
+@@ -31,7 +31,7 @@ contract UUPSUpgradeableLegacyMock is UUPSUpgradeableMock {
+         if (!rollbackTesting.value) {
+             // Trigger rollback using upgradeTo from the new implementation
+             rollbackTesting.value = true;
+-            Address.functionDelegateCall(
++            __functionDelegateCall(
+                 newImplementation,
+                 abi.encodeWithSignature("upgradeTo(address)", oldImplementation)
+             );
+@@ -51,4 +51,14 @@ contract UUPSUpgradeableLegacyMock is UUPSUpgradeableMock {
+     function upgradeToAndCall(address newImplementation, bytes memory data) public payable override {
+         _upgradeToAndCallSecureLegacyV1(newImplementation, data, false);
+     }
++
++    // ERC1967Upgrade._functionDelegateCall is private so we reproduce it here.
++    // An extra underscore prevents a name clash error.
++    function __functionDelegateCall(address target, bytes memory data) private returns (bytes memory) {
++        require(Address.isContract(target), "Address: delegate call to non-contract");
++
++        // solhint-disable-next-line avoid-low-level-calls
++        (bool success, bytes memory returndata) = target.delegatecall(data);
++        return Address.verifyCallResult(success, returndata, "Address: low-level delegate call failed");
++    }
+ }
```
same

Amxx commented on scripts/upgradeable/upgradeable.patch:
```diff
@@ -0,0 +1,350 @@
+   },
+   "homepage": "https://openzeppelin.com/contracts/"
+ }
+diff --git a/contracts/proxy/ERC1967/ERC1967Upgrade.sol b/contracts/proxy/ERC1967/ERC1967Upgrade.sol
+index 0680f354..ebd83b61 100644
+--- a/contracts/proxy/ERC1967/ERC1967Upgrade.sol
++++ b/contracts/proxy/ERC1967/ERC1967Upgrade.sol
+@@ -63,7 +63,7 @@ abstract contract ERC1967Upgrade {
+     function _upgradeToAndCall(address newImplementation, bytes memory data, bool forceCall) internal {
+         _upgradeTo(newImplementation);
+         if (data.length > 0 || forceCall) {
+-            Address.functionDelegateCall(newImplementation, data);
++            _functionDelegateCall(newImplementation, data);
+         }
+     }
+ 
+@@ -165,7 +165,21 @@ abstract contract ERC1967Upgrade {
+         _setBeacon(newBeacon);
+         emit BeaconUpgraded(newBeacon);
+         if (data.length > 0 || forceCall) {
+-            Address.functionDelegateCall(IBeacon(newBeacon).implementation(), data);
++            _functionDelegateCall(IBeacon(newBeacon).implementation(), data);
+         }
+     }
++
++    /**
++     * @dev Same as {xref-Address-functionCall-address-bytes-string-}[`functionCall`],
++     * but performing a delegate call.
++     *
++     * _Available since v3.4._
++     */
++    function _functionDelegateCall(address target, bytes memory data) private returns (bytes memory) {
++        require(Address.isContract(target), "Address: delegate call to non-contract");
++
++        // solhint-disable-next-line avoid-low-level-calls
++        (bool success, bytes memory returndata) = target.delegatecall(data);
++        return Address.verifyCallResult(success, returndata, "Address: low-level delegate call failed");
++    }
+ }
```
same

Amxx commented on scripts/upgradeable/upgradeable.patch:
```diff
@@ -0,0 +1,350 @@
+  */
+ contract TokenTimelock {
+     using SafeERC20 for IERC20;
+diff --git a/contracts/utils/Address.sol b/contracts/utils/Address.sol
+index 433a866d..64a80d35 100644
+--- a/contracts/utils/Address.sol
++++ b/contracts/utils/Address.sol
+@@ -161,31 +161,6 @@ library Address {
+         return verifyCallResultFromTarget(target, success, returndata, errorMessage);
+     }
+ 
+-    /**
+-     * @dev Same as {xref-Address-functionCall-address-bytes-}[`functionCall`],
+-     * but performing a delegate call.
+-     *
+-     * _Available since v3.4._
+-     */
+-    function functionDelegateCall(address target, bytes memory data) internal returns (bytes memory) {
+-        return functionDelegateCall(target, data, "Address: low-level delegate call failed");
+-    }
+-
+-    /**
+-     * @dev Same as {xref-Address-functionCall-address-bytes-string-}[`functionCall`],
+-     * but performing a delegate call.
+-     *
+-     * _Available since v3.4._
+-     */
+-    function functionDelegateCall(
+-        address target,
+-        bytes memory data,
+-        string memory errorMessage
+-    ) internal returns (bytes memory) {
+-        (bool success, bytes memory returndata) = target.delegatecall(data);
+-        return verifyCallResultFromTarget(target, success, returndata, errorMessage);
+-    }
+-
+     /**
+      * @dev Tool to verify that a low level call to smart-contract was successful, and revert (either by bubbling
+      * the revert reason or using the provided one) in case of unsuccessful call or if target was not a contract.
```
should no longer be removed

Amxx commented on scripts/upgradeable/upgradeable.patch:
```diff
@@ -0,0 +1,350 @@
+     /**
+      * @dev Tool to verify that a low level call to smart-contract was successful, and revert (either by bubbling
+      * the revert reason or using the provided one) in case of unsuccessful call or if target was not a contract.
+diff --git a/contracts/utils/Multicall.sol b/contracts/utils/Multicall.sol
+index 7470c559..9bcb8605 100644
+--- a/contracts/utils/Multicall.sol
++++ b/contracts/utils/Multicall.sol
+@@ -18,8 +18,22 @@ abstract contract Multicall {
+     function multicall(bytes[] calldata data) external virtual returns (bytes[] memory results) {
+         results = new bytes[](data.length);
+         for (uint256 i = 0; i < data.length; i++) {
+-            results[i] = Address.functionDelegateCall(address(this), data[i]);
++            results[i] = _functionDelegateCall(address(this), data[i]);
+         }
+         return results;
+     }
++
++    /**
++     * @dev Same as {xref-Address-functionCall-address-bytes-string-}[`functionCall`],
++     * but performing a delegate call.
++     *
++     * _Available since v3.4._
++     */
++    function _functionDelegateCall(address target, bytes memory data) private returns (bytes memory) {
++        require(Address.isContract(target), "Address: delegate call to non-contract");
++
++        // solhint-disable-next-line avoid-low-level-calls
++        (bool success, bytes memory returndata) = target.delegatecall(data);
++        return Address.verifyCallResult(success, returndata, "Address: low-level delegate call failed");
++    }
+ }
```
same

Amxx reviewed (commented):
patch could be simplified

Amxx commented on hardhat/env-artifacts.js:
```diff
@@ -0,0 +1,20 @@
+const { HardhatError } = require('hardhat/internal/core/errors');
+
+extendEnvironment(env => {
```
IMO, there should be a comment explaining what this actually does and why its needed.

Amxx commented on hardhat/task-test-get-files.js:
```diff
@@ -0,0 +1,36 @@
+const { internalTask } = require('hardhat/config');
+const { TASK_TEST_GET_TEST_FILES } = require('hardhat/builtin-tasks/task-names');
+
+internalTask(TASK_TEST_GET_TEST_FILES).setAction(async ({ testFiles }, { config }) => {
```
same, we need some documentation here

frangio replied:
There was a comment at the top of the file. But I moved it down and rewrote it a little.

Amxx commented on scripts/upgradeable/patch-apply.sh:
```diff
@@ -0,0 +1,19 @@
+set -euo pipefail
+
+DIRNAME="$(dirname -- "${BASH_SOURCE[0]}")"
+PATCH="$DIRNAME/upgradeable.patch"
```
I'm wondering if we should have multiple patch files:
```$DIRNAME/upgradeable-patches/*.patch```

The advantage is that we could more effectivelly segment the different things that are patched (instead of having a single difficult to read patch file). The downside is that patch generation is more complexe. Maybe that is something we can delay until we actually need it (if ever)

frangio replied:
Yeah the reason for a single patch is that `patch-save.sh` can be super simple.

Hasinafidahussain reviewed (commented):
- 

> #Sir easy pisa ya bank ka account number 03443616250 per Apne name k  sath send kare 

>

frangio commented on scripts/upgradeable/transpile-onto.sh:
I'm sorry that this file has no comments... I wrote it a long time ago and I should've written the comments back then. :grimacing:

frangio commented:
Moved this back into draft because after merging the latest `master` I found an issue with the EIP712 tests caused by the transpiler.

frangio commented:
I addressed the EIP712 issues and this should be good now. It turned out to be tricky so here are some notes to help understand.

The EIP712 contract uses immutable variables quite extensively, and it's even designed to work well behind a proxy. I think we initially intended that code to go as-is in the upgradeable contracts, however, once we transpile the rest of the codebase we run into conflicts because those immutable variables need to be set in a constructor whereas everything else will be changed to initializers, and the transpiler doesn't support this kind of hybrid transformation.

To fix this, the [upgradeable patch](https://github.com/frangio/openzeppelin-contracts/blob/transpile/scripts/upgradeable/upgradeable.patch#L272) now removes most of the immutable variables (the remaining are converted into storage variables by the transpiler), including the cached domain separator. The new code after transpilation can be seen in [this branch in my fork](https://github.com/frangio/openzeppelin-contracts/blob/transpile-result/contracts/utils/cryptography/EIP712Upgradeable.sol), and is very similar to the [`EIP712`](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/release-v4.8/contracts/utils/cryptography/EIP712Upgradeable.sol) we have in v4.8.

One of the changes in v4.9 is that we now need the name and version for `eip712Domain()`, so some internal functions that were previously added exclusively in the upgradeable variant had to be refactored: instead of overrideable [`_EIP712NameHash` and `_EIP712VersionHash`](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/release-v4.8/contracts/utils/cryptography/EIP712Upgradeable.sol#L95-L113) we now have overrideable [`_EIP712Name` and `_EIP712Version`](https://github.com/frangio/openzeppelin-contracts/blob/transpile-result/contracts/utils/cryptography/EIP712Upgradeable.sol#L137-L155).

Another addition I made, that may be controversial, is that since there are new storage variables (name and version) that have to be initialized, I added code to deal with an upgraded but accidentally uninitialized contract. [`_EIP712NameHash` and `_EIP712VersionHash`](https://github.com/frangio/openzeppelin-contracts/blob/transpile-result/contracts/utils/cryptography/EIP712Upgradeable.sol#L157-L197) will read from the old storage variables if it detects that the new ones were not initialized, and `eip712Domain()` will revert if accurate name and version values can't be returned due to uninitialized storage.

ernestognw commented on scripts/upgradeable/README.md:
```diff
@@ -0,0 +1,21 @@
+
+## `transpile.sh`
+
+Applies patches and invokes the transpiler with the necessary command line flags.
```
Feel free to adapt this suggestion as you wish, I just feel it's not explicit enough to just say `the necessary contracts`
```suggestion
Applies patches and invokes the transpiler with necessary command line flags to include or exclude certain purpose-specific contracts
```

ernestognw replied:
This is fine, I also don't think we need to be exhaustive but the original writing was not saying much 😅 thanks!

ernestognw commented on scripts/upgradeable/transpile-onto.sh:
```diff
@@ -0,0 +1,38 @@
+  git checkout "$target"
+else
+  git checkout --orphan "$target"
+  if [ -n "$base" ] && git rev-parse -q --verify "$base"; then
+    git reset --soft "$base"
+  fi
```
I have the feeling that we should add an `else` here if there's no `base` and `target` doesn't exist. My concern is that the next check will fail because the transpiler runs `patch-apply.sh`, adding changes:

```bash
if ! git diff --quiet --cached; then
  git commit -m "Transpile $commit"
fi
```

frangio replied:
I'm not sure I understand the comment/concern.

I've pushed a commit adding some comments in this file. Let me know if those clear things up.

ernestognw replied:
My concern is that the transpile will add changes even if there's no `target` or `base`, in other words, there's no `else` case in which we fail (`exit 1`) from the script. In such case, we'll commit without the actual transpilation and then push to the `contracts-upgradeable` with a commit not including the upgradeable contracts.

Note that it'll only happen if we don't provide `target` or `base`, which can happen due to a bug or human error changing the scripts. Right now it works so I'll suggest to just add a comment, but I wouldn't block the merge since we know it works as of now.

frangio commented on test/utils/cryptography/EIP712.test.js:
```diff
@@ -60,11 +60,11 @@ contract('EIP712', function (accounts) {
-            const expectedSeparator = await domainSeparator(cloneDomain);
-            expect(await clone.$_domainSeparatorV4()).to.equal(expectedSeparator);
-
             const reportedDomain = await getDomain(clone);
             expect(mapValues(reportedDomain, String)).to.be.deep.equal(mapValues(cloneDomain, String));
+
+            const expectedSeparator = await domainSeparator(cloneDomain);
+            expect(await clone.$_domainSeparatorV4()).to.equal(expectedSeparator);
```
Previously the first assertion was about the domain separator being as expected. It's a hash, so seeing it fail is very opaque, just that a bytes32 value is not the one you wanted. I moved that assertion second, and the first one now compares the domain fields in cleartext so the errors are more understandable. It's only related to this PR in that I ran into this difficulty while working on the EIP712 upgradeability changes.

ernestognw replied:
The rationale of letting a more clear error to show up first is enough. Thanks!

frangio commented on contracts/utils/cryptography/EIP712.sol:
```diff
@@ -30,7 +30,7 @@ import "../../interfaces/IERC5267.sol";
  * _Available since v3.4._
  *
- * @custom:oz-upgrades-unsafe-allow state-variable-immutable state-variable-assignment
+ * @custom:oz-upgrades-unsafe-allow state-variable-immutable state-variable-assignment constructor
```
I should remove this, I think I added it in a previous iteration.

frangio commented on scripts/upgradeable/transpile.sh:
```diff
@@ -0,0 +1,35 @@
+# -D: delete original and excluded files
+# -b: use this build info file
+# -i: use included Initializable
+# -x: exclude all proxy contracts except Clones library
```
This comment is outdated. I think it was Clones first and afterwards we added UUPS, etc. I've updated the comment.

frangio commented on scripts/upgradeable/README.md:
````diff
@@ -0,0 +1,21 @@
+## `transpile-onto.sh`
+
+```
+bash scripts/upgradeable/transpile-onto.sh <target> [base]
````
I was confused. Checking [docopt](http://docopt.org/) for reference and it seems to use `[<base>]` as you suggest. Changed.

frangio commented on hardhat/task-test-get-files.js:
```diff
@@ -0,0 +1,36 @@
+          'proxy/beacon/UpgradeableBeacon.test.js',
+          'proxy/ERC1967/ERC1967Proxy.test.js',
+          'proxy/transparent/ProxyAdmin.test.js',
+          'proxy/transparent/TransparentUpgradeableProxy.behaviour.js',
```
I'm not sure why I excluded this initially but it doesn't seem to be necessary.

ernestognw commented on scripts/upgradeable/README.md:
````diff
@@ -0,0 +1,21 @@
+bash scripts/upgradeable/transpile-onto.sh <target> [<base>]
+```
+
+Transpiles the contents of the current git branch and commits the result as a new commit on branch `<target>`. If branch `<target>` doesn't exist, it will copy the commit history of `[base]` (this is used in GitHub Actions, but is usually not necessary locally).
````
```suggestion
Transpiles the contents of the current git branch and commits the result as a new commit on branch `<target>`. If branch `<target>` doesn't exist, it will copy the commit history of `[<base>]` (this is used in GitHub Actions, but is usually not necessary locally).
```

ernestognw reviewed (approved):
Just a nit, but feel free to merge if accepting the suggestion dismisses my review

ernestognw reviewed (approved):
Just a nit, but feel free to merge if accepting the suggestion dismisses my review

Amxx reviewed (commented):
that is a lot to review, I'm about 50% done

Amxx commented on scripts/upgradeable/upgradeable.patch:
```diff
@@ -0,0 +1,481 @@
+         );
+     }
++
++    /**
++     * @dev The name parameter for the EIP712 domain.
++     *
++     * NOTE: This function reads from storage by default, but can be redefined to return a constant value if gas costs
++     * are a concern.
++     */
++    function _EIP712Name() internal virtual view returns (string memory) {
++        return _name;
++    }
++
++    /**
++     * @dev The version parameter for the EIP712 domain.
++     *
++     * NOTE: This function reads from storage by default, but can be redefined to return a constant value if gas costs
++     * are a concern.
++     */
++    function _EIP712Version() internal virtual view returns (string memory) {
++        return _version;
++    }
```
IMO this part should be in the vanilla contract. That would reduce the patch size, and improve "compatibility" between both versions.

(or be private?)

Amxx replied:
The argument that we don't want to ship that with the "vanila" version without having it audited doesn't really hold if we and up shipping it with the upgradeable version (without it going through the audit)

Amxx reviewed (commented):
I'll @frangio and @ernestognw process these and merge with my implicit approval that they are going to do things right.

frangio commented on .github/workflows/upgradeable.yml:
```diff
@@ -7,17 +7,24 @@ on:
+          repository: OpenZeppelin/openzeppelin-contracts-upgradeable
+          fetch-depth: 0
+          token: ${{ secrets.GH_TOKEN_UPGRADEABLE }}
+      - name: Fetch current non-upgradeable branch
+        run: |
+          git fetch "https://github.com/${{ github.repository }}.git" "$REF"
+          git checkout FETCH_HEAD
+        env:
+          REF: ${{ github.ref }}
```
The reason for doing it this way is to avoid command injection. I believe `github.ref` is potentially attacker controlled so if we just interpolate it into the command like you suggest it could result in command injection. Usign an environment variable prevents this attack vector.

`github.repository` is known so we don't need to do the same thing.

See https://securitylab.github.com/research/github-actions-untrusted-input for more about this.

frangio commented on test/utils/Checkpoints.test.js:
```diff
@@ -6,137 +6,161 @@ const { batchInBlock } = require('../helpers/txpool');
-            expect(await getAtBlockVariant(this.mock, this.tx1.receipt.blockNumber - 1)).to.be.bignumber.equal('0');
-            expect(await getAtBlockVariant(this.mock, this.tx1.receipt.blockNumber)).to.be.bignumber.equal('1');
-            expect(await getAtBlockVariant(this.mock, this.tx2.receipt.blockNumber)).to.be.bignumber.equal('2');
+            expect(
+              await this.methods.history[getAtBlockVariant](this.tx1.receipt.blockNumber - 1),
+            ).to.be.bignumber.equal('0');
+            expect(await this.methods.history[getAtBlockVariant](this.tx1.receipt.blockNumber)).to.be.bignumber.equal(
+              '1',
+            );
+            expect(await this.methods.history[getAtBlockVariant](this.tx2.receipt.blockNumber)).to.be.bignumber.equal(
+              '2',
+            );
```
Yeah, it's the linter... I tried a few things but I don't think the results are better.
